internal/admin/admin.go

```diff
@@ -163,17 +165,57 @@ func (adm *Admin) Metadata(ctx context.Context) (*varlogpb.MetadataDescriptor, e
 }
 
 func (adm *Admin) getStorageNode(ctx context.Context, snid types.StorageNodeID) (*vmspb.StorageNodeMetadata, error) {
-	snm, ok := adm.statRepository.GetStorageNode(snid)
-	if !ok {
-		return nil, errors.WithMessagef(admerrors.ErrNoSuchStorageNode, "get storage node %d", int32(snid))
+	adm.mu.RLock()
+	defer adm.mu.RUnlock()
+
+	md, err := adm.mrmgr.ClusterMetadataView().ClusterMetadata(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Unavailable, "get storage node: cluster metadata not fetched")
+	}
+	snd := md.GetStorageNode(snid)
+	if snd == nil {
+		return nil, status.Errorf(codes.NotFound, "get storage node: %d", int32(snid))
+	}
+	if snm, ok := adm.statRepository.GetStorageNode(snid); ok {
+		return snm, nil
 	}
-	return snm, nil
+	return &vmspb.StorageNodeMetadata{
+		StorageNodeMetadataDescriptor: snpb.StorageNodeMetadataDescriptor{
+			ClusterID:   adm.cid,
+			StorageNode: snd.StorageNode,
+		},
+		CreateTime: snd.CreateTime,
+	}, nil
 }
 
 func (adm *Admin) listStorageNodes(ctx context.Context) ([]vmspb.StorageNodeMetadata, error) {
 	adm.mu.RLock()
 	defer adm.mu.RUnlock()
-	return adm.statRepository.ListStorageNodes(), nil
+
+	md, err := adm.mrmgr.ClusterMetadataView().ClusterMetadata(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Unavailable, "list storage nodes: cluster metadata not fetched")
+	}
+
+	snms := make([]vmspb.StorageNodeMetadata, 0, len(md.StorageNodes))
+	snmsMap := adm.statRepository.ListStorageNodes()
+	for _, snd := range md.StorageNodes {
+		if snm, ok := snmsMap[snd.StorageNodeID]; ok {
+			snms = append(snms, *snm)
+			continue
+		}
+		snms = append(snms, vmspb.StorageNodeMetadata{
+			StorageNodeMetadataDescriptor: snpb.StorageNodeMetadataDescriptor{
```

Comment:
Is it ok to not have the Storages metadata?

Reply:
Zero values of some fields indicate that the admin does not check the heartbeat of the storage node yet.
Clients such as KOV should check whether they, for instance, LastHeartbeatTime are zero.

Reply:
`Admin.GetStorageNodes` uses `ListStorageNodes`.  So, there will be a lot of difference from before.
How about returning Unavailable if statRepository doesn't have all the metadata?

Reply:
> `Admin.GetStorageNodes` uses `ListStorageNodes`. So, there will be a lot of difference from before. 

Originally, `GetStorageNodes` has been implemented by `ListStorageNodes`. So, it has been deprecated. Don't worry, you can use `GetStorageNodes` still.

> How about returning Unavailable if statRepository doesn't have all the metadata?

ListStorageNodes should be used by the storage node's operational script - `start_varlogsn.py`. For the storage node to decide whether it is registered to the cluster, ListStorageNodes must return all registered storage nodes regardless of reachability.

If the clients should know that the result of ListStorageNodes is partial, that is, some storage nodes are outage, it can check whether some of the result lists have zero value for LastHeartbeatTime.
However, if you think it is very cumbersome, please tell me. We are open to change or other approaches.

Reply:
What I mean is that kovadm works under the assumption that varlogadm always gives all metadata.
Okay. We will change the metadata processing logic in kovadm.